File: LAB 1/BFS.cpp
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

class Solution
{
public:
    vector<int> BFS(int numberOfVertices, vector<int> adjacencyList[])
    {
        int visited[numberOfVertices] = {0}; // 0 means not visited, 1 means visited.
        visited[0] = 1;                      // mark the first node as visited.

        queue<int> q;
        q.push(0);
        vector<int> result;

        // iteratation till queue is empty.
        while (!q.empty())
        {
            int node = q.front();   // get the front of the queue and store it in a variable called "node".
            q.pop();                // pop the front of the queue.
            result.push_back(node); // push the node into the result vector.

            for (auto it : adjacencyList[node])
            {
                // if the node is not visited, push it into the queue and mark it as visited.
                if (!visited[it])
                {
                    q.push(it);
                    visited[it] = 1;
                }
            }
        };
        return result;
    };
};

void addingAdjacency(vector<int> adj[], int u, int v)
{
    adj[u].push_back(v);
    adj[v].push_back(u);
}

void print(vector<int> &ans)
{
    for (int i = 0; i < ans.size(); i++)
    {
        cout << ans[i] << " ";
    }
}

int main()
{
    // vector<int> adj[6];

    // addingAdjacency(adj, 0, 1);
    // addingAdjacency(adj, 1, 2);
    // addingAdjacency(adj, 1, 3);
    // addingAdjacency(adj, 0, 4);

    // Solution obj;
    // vector<int> ans = obj.BFS(5, adj);
    // print(ans);

    int numberOfVertices, numberOfEdges;

    cout << "Enter the number of vertices: ";
    cin >> numberOfVertices;

    cout << "Enter the number of edges: ";
    cin >> numberOfEdges;

    vector<int> adj[numberOfVertices];
    
    cout << "Enter the edges (u v):" << endl;
    for (int i = 0; i < numberOfEdges; i++)
    {
        int u, v;
        cin >> u >> v;
        addingAdjacency(adj, u, v);
    }

    Solution obj;
    vector<int> ans = obj.BFS(numberOfVertices, adj);

    cout << "BFS Traversal: ";
    print(ans);

    return 0;
}

File: LAB 1/BFS.py
from collections import defaultdict

class Graph:
    def __init__(self):
        self.graph = defaultdict(list)
    
    def addEdge(self, u, v):
        self.graph[u].append(v)
    
    def BFS(self, s):
        # Create a visited list with size based on the maximum node in the graph
        visited = [False] * (max(self.graph) + 1)
        queue = []  # Initialize a queue

        queue.append(s)  # Enqueue the starting vertex
        visited[s] = True  # Mark the starting vertex as visited

        while queue:  # Loop until the queue is empty
            s = queue.pop(0)  # Dequeue the front of the queue
            print(s, end=" ")  # Print the current node

            # Iterate through all neighbors of the current node
            for i in self.graph[s]:
                if not visited[i]:  # If the neighbor is unvisited
                    queue.append(i)  # Enqueue the neighbor
                    visited[i] = True  # Mark the neighbor as visited

if __name__ == '__main__':
    g = Graph()
    
    # Input number of edges and edges themselves
    num_edges = int(input("Enter the number of edges: "))
    print("Enter the edges (u v):")
    for _ in range(num_edges):
        u, v = map(int, input().split())
        g.addEdge(u, v)

    # Input starting vertex
    start_vertex = int(input("Enter the starting vertex for BFS: "))

    print("Following is Breadth First Traversal:")
    g.BFS(start_vertex)
